packages/lesswrong/components/posts/TruncatedAuthorsList.tsx

```diff
@@ -1,114 +1,179 @@
-    setOverflowCount(0);
+  if (!ref.current || !expandContainer.current || isAnon) {
     return;
   }
-  const afterWidth = afterRef.current.getBoundingClientRect().width ?? 0;
-  const containerRight = ref.current.getBoundingClientRect().right;
-  const elems: NodeListOf<HTMLSpanElement> = ref.current.querySelectorAll(selector);
-  elems[elems.length - 1].after(afterRef.current);
-  for (let i = 0; i < elems.length; ++i) {
-    elems[i].style.opacity = "100%";
+
+  // Get the 'off-screen' scratch buffer (it's actually on-screen with opacity 0)
+  const scratch = ref.current.querySelector("." + classes.scratch);
+  if (!scratch) {
+    return;
   }
-  for (let i = 1; i < elems.length; ++i) {
-    const elemBounds = elems[i].getBoundingClientRect();
-    if (elemBounds.right + afterWidth > containerRight) {
-      setOverflowCount(elems.length - i);
-      elems[i].before(afterRef.current);
-      for ( ; i < elems.length; ++i) {
-        elems[i].style.opacity = "0%";
-      }
+
+  // Find the '+ n more' node and make sure it's at the end of the scratch node
+  let more = scratch.querySelector("." + classes.more);
+  if (!more) {
+    more = ref.current.querySelector("." + classes.more);
+    if (!more) {
       return;
     }
+    scratch.appendChild(more);
   }
-  setOverflowCount(0);
+
+  // Remove the placeholder if it exists
+  const placeholder = ref.current.querySelector("." + classes.placeholder);
+  if (placeholder) {
+    ref.current.removeChild(placeholder);
+  }
+
+  // Move all the authors into the scratch node
+  const displayedAuthors = ref.current.querySelectorAll("." + classes.item);
+  for (const author of Array.from(displayedAuthors).reverse()) {
+    scratch.insertBefore(author, scratch.firstChild);
+  }
+
+  // Find how much space we have and what needs to fit there
+  const maxWidth = getMaxExpandableWidth(expandContainer.current);
+  const authors = Array.from(scratch.querySelectorAll("." + classes.item));
+  const bounds = authors.map((author) => author.getBoundingClientRect());
+
+  // Calculate how may authors we can fit
+  let width = bounds[0].width;
+  let moreCount = 0;
+  let i = 1;
+  for ( ; i < bounds.length; ++i) {
+    const newWidth = width + bounds[i].width + TRUNCATION_PADDING;
+    if (newWidth > maxWidth) {
+      const moreWidth = more?.getBoundingClientRect().width;
+      if (width + moreWidth > maxWidth) {
+        i--;
+      }
+      moreCount = bounds.length - i;
+      break;
+    }
+    width = newWidth;
+  }
+
+  // Move all the visible authors to the right place and set n in '+ n more' if shown
+  ref.current.innerHTML = "";
+  for (let j = 0; j < i; ++j) {
+    ref.current.appendChild(authors[j]);
+  }
+  if (moreCount) {
+    const text = moreCount === authors.length
+      ? `${moreCount} authors`
+      : `+ ${moreCount} more`
+    more.innerHTML = more.innerHTML.replace(/(\+ \d+ more)|(\d+ authors)/, text);
+    ref.current.appendChild(more);
```

Comment:
How much faster is it to do this direct dom manipulation that waiting for the react refresh? You're stuck waiting for the JS initialization either way, and for the rehydration to reach this step to trigger the useEffect. (Genuinely curious, probably don't hold up shipping to answer though sadly.)

Reply:
Without having measured, I'd guess pretty darn fast simply because we're not creating/destroying any DOM nodes, just moving them around to different parents. (It also just makes the code simpler - this kind of thing is pretty annoying in React).

There's also a rare-but-possible worst case scenario in React where the modified version causes a layout change that causes it to swith back to the original version, resulting in an infitite rerender loop.